### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by Ney2**
This is a PR for the  RoR final capstone project. In this project I have achieved the  following requirements :-
- Set up the repository and tools.
- Create your models according to the ERD diagram. Don't forget about validations and tests.
- Implement the authentication and authorization.

- Create the page to add a category.
- Create the home page.
- Create the page to add a transaction.
- Create the transactions page.

- Make sure your app is tested adequately.
- Deploy the project, and test for final details.
- Record a video for your project.
- Create a good README and PR description.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,73 +1,46 @@
...
+### Show your support
 Give a ⭐️ if you like this project!
 
-## Acknowledgments
-
-- Hat tip to anyone whose code was used
-- Inspiration
-- etc
+### 🤝 Contributing
+Contributions, issues, and feature requests are welcome! Feel free to check the issues page.
```
Comment (BenMukebo) [inline README.md]: - [ ] I would also recommend adding the issue link in the contributing sections: 

```
## 🤝 Contributing

Contributions, issues, and feature requests are welcome!

Feel free to check the [issues page](../../issues/).
```

File: app/views/devise/registrations/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      <ion-icon name="key-outline"></ion-icon>
+      <%= f.password_field :password, autocomplete: "new-password", class: 'option-text', placeholder: 'Password' %>
+    </div>
+    <div class="user__option category-title">
+      <ion-icon name="lock-closed-outline"></ion-icon>
+      <%= f.password_field :password, autocomplete: "new-password", class: 'option-text', placeholder: 'Confirm Password' %>
+    </div>
+  </div>
+  <%= f.button 'Sign Up', class: 'user__save' %><% end %> <%= render "devise/shared/links" %>
+</section>
```
Comment (BenMukebo) [inline app/views/devise/registrations/new.html.erb:28]: - [ ] Your app working fine but you should stick to the design as required font, colors, layout, and ... as in the requirements. 🤔  I will suggest you add the missing navigation bar with the back button (<) to the splash page. 😉 
- Please add this to the login page as well 👍🏻 
- Design 
Screenshot 2022-11-17 at 13 13 48

File: app/views/groups/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        <% end %>
+      </li>
+    </ul>
+  </div>
+  <div class="options">
+    <%= button_to 'Log out', destroy_user_session_path, method: :delete, class: 'btn btn-danger'%>
+    <%= link_to new_group_path do %>
+    <ion-icon name="add-outline" class="add-button">Add</ion-icon>
+    <% end %>
+    <ion-icon name="alert-circle" class="show-button" data-toggle="modal" data-target="#exampleModal"></ion-icon>
```
Comment (BenMukebo) [inline app/views/groups/index.html.erb]: - [ ] Your app working fine but you should stick to the design as required font, colors, layout and space elements as in the requirements 

|- Design  |  - Yours  |
| ---------:|:---------|
|Screenshot 2022-11-17 at 13 46 17 | budget-app |

-  let's improve the design of the "add a new category"   button at the bottom that brings the user to the page to create a new category to match the original disign eg: paddibg, margnin, color, background-color, ....

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+  has_many :groups, dependent: :destroy
+  has_many :entities, dependent: :destroy
+  has_one_attached :avatar, dependent: :destroy
+  validates :name, presence: true, uniqueness: true
```
Comment (BenMukebo) [inline app/models/user.rb]: - [ ] Great job using the devise for authentication however you need to validate all user input  (email, password) to make sure that anyone with bad intentions cannot compromise their app. Please add this 👍🏻

File: spec/integrations/group_spec.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+RSpec.feature 'Groups', type: :feature do
+  background do
+    visit new_user_registration_path
+    fill_in 'Name', with: 'Kidist'
+    fill_in 'Email', with: 'kidist@gmail.com'
+    fill_in 'Password', with: '1234567'
+    fill_in 'Confirm Password', with: '1234567'
+    click_button 'Sign Up'
+
```
Comment (BenMukebo) [inline spec/integrations/group_spec.rb:11]: - [ ] Great so far 👏🏻 However you should write integration tests for all the most important components of the application. Kindly make sure to add integration tests for other important components. ✍🏻

File: app/models/group.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Group < ApplicationRecord
+  has_many :entities, dependent: :destroy
```
Comment (BenMukebo) [inline app/models/group.rb]: - [ ] 🤔 Since the user can have many Groups Why don't you add `belongs_to :user` into the Group Model?

**Review by BenMukebo — CHANGES_REQUESTED** [review]
## Hi @Ney2, 

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


### To Highlight 🎉✔️ 
- Git flow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️
- Good presentation video. ✔️ 
- The database schema reflect the Entity Relationship Disgram 👍🏻 
- The app has Splash screen with links to the sign up and log in pages 👍🏻 
- Your Design looks awesome, keep rocking :rocket: 

## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_

- Note: Most of the time as developers, we will receive loosely defined business requirements from the system analysts/product managers, and we must learn skills to create minimum viable products from them.  In real life after you get a job that is what will you do, _code based on the design that is given to you_. If we focus on things outside of the requirements or if we decide to change part of the requirement, this can affect our productivity and will look bad as a team player. I like your design as well but it is very important that your design be closed to the required one.  

- Screenshot of the requirements 
Screenshot 2022-11-17 at 13 13 06

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers :clinking_glasses:  and Happy coding!!! :dancers:💻  

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 1 above)
Comment (Ney2) [reply README.md]: Solved thank you

File: app/models/group.rb (hunk 6 above)
Comment (Ney2) [reply app/models/group.rb]: Solved thank you

File: app/models/user.rb (hunk 4 above)
Comment (Ney2) [reply app/models/user.rb]: Solved thank you

File: app/views/devise/registrations/new.html.erb (hunk 2 above)
Comment (Ney2) [reply app/views/devise/registrations/new.html.erb:28]: Solved thank you

File: spec/integrations/group_spec.rb (hunk 5 above)
Comment (Ney2) [reply spec/integrations/group_spec.rb:11]: Solved thank you

File: app/views/groups/index.html.erb (hunk 3 above)
Comment (Ney2) [reply app/views/groups/index.html.erb]: Solved thank you

**Comment by Ney2** [comment]
- I have implemented all the changes requested by the previous reviewer

splash
login
signup
transaction
category
entities
newentity

**Review by FrederickMih — APPROVED** [review]
Hi @Ney2, 

In my opinion, your project is ready for presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

## Highlights :green_circle: 

- Associations between models have been correctly set up :+1: 
- App has unit and integration tests and all tests ran without errors :+1: 
- PR is professional :+1: 

Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Comment by Ney2** [comment]
Thank you for reviewing and approving my project

**Pull request merged**